### PR TITLE
Memory bank distributed training support

### PR DESCRIPTION
**Pull request by guarin**
## Changes

* Add feature dimension when setting memory bank size. This is required for distributed training because the buffer size cannot be changed dynamically.
* Cleanup memory bank implementation.
* Add `feature_dim_first` argument to memory bank. This is required for backwards compatibility. In the future we should be able to set `feature_dim_first=False` and remove the requirement for transposing features coming from the feature bank.
* Update classes/functions using memory bank  from `size=num_features` to `size=(num_features, dim)`.

## How was it tested?
* Added unit tests

### CODE REVIEW

File: benchmarks/imagenet/resnet50/main.py [hunk 1]
```diff
@@ -230,7 +230,7 @@ def pretrain(
         logger=TensorBoardLogger(save_dir=str(log_dir), name="pretrain"),
         precision=precision,
         strategy="ddp_find_unused_parameters_true",
-        sync_batchnorm=True,
+        sync_batchnorm=accelerator != "cpu",  # Sync batchnorm is not supported on CPU.
```
Comment (guarin) [inline benchmarks/imagenet/resnet50/main.py:233]: Added this to make benchmarks runnable on CPU, makes testing distributed training easier.

File: lightly/models/modules/memory_bank.py [hunk 2]
```diff
@@ -20,26 +22,30 @@ class MemoryBankModule(torch.nn.Module):
...
+            dimension is inferred from the first batch stored in the memory bank.
+            Leaving out the feature dimension might lead to errors in distributed
+            training.
         gather_distributed:
             If True then negatives from all gpus are gathered before the memory bank
             is updated. This results in more frequent updates of the memory bank and
             keeps the memory bank contents independent of the number of gpus. But it has
             the drawback that synchronization between processes is required and
             diversity of the memory bank content is reduced.
+        feature_dim_first:
```
Comment (guarin) [inline lightly/models/modules/memory_bank.py:37]: `feature_dim_first=False` makes working with memory bank much easier.

File: lightly/utils/benchmarking/metric_callback.py [hunk 3]
```diff
@@ -60,7 +60,7 @@ def _append_metrics(
         self, metrics_dict: Dict[str, List[float]], trainer: Trainer
     ) -> None:
         for name, value in trainer.callback_metrics.items():
-            if isinstance(value, float) or (
-                isinstance(value, Tensor) and value.numel() == 1
-            ):
-                metrics_dict.setdefault(name, []).append(float(value))
+            if isinstance(value, Tensor) and value.numel() != 1:
+                # Skip non-scalar tensors.
+                continue
```
Comment (guarin) [inline lightly/utils/benchmarking/metric_callback.py:65]: `trainer.callback_metrics.items()`  returns `(str, Tensor)` tuples and mypy pointed out that `isinstance(value, float)` is impossible if the value is always a tensor. As this check previously passed CI, I believe that either pytorch lightning changed the typing of this function, or I have a different mypy version locally. In any case, I changed the test slightly to make it pass in all cases. The functionality should remain exactly the same :)

File: lightly/models/modules/memory_bank.py [hunk 4]
```diff
@@ -48,38 +54,67 @@ class MemoryBankModule(torch.nn.Module):
...
+                ),
+                UserWarning,
+            )
+        else:
+            self._init_memory_bank(dim=None)
+
     @torch.no_grad()
-    def _init_memory_bank(self, dim: int) -> None:
-        """Initialize the memory bank if it's empty
+    def _init_memory_bank(self, dim: Union[Sequence[int], None]) -> None:
```
Comment (philippmwirth) [inline lightly/models/modules/memory_bank.py:101]: This is a bit convoluted I think it'd be a better solution to pass the `size` explicitly:
```python
    @torch.no_grad()
    def _init_memory_bank(self, size: Sequence[int]) -> None:
        """Initialize the memory bank if it's empty.
        Args:
            size:
                The dimensions of the memory bank.
        """
        self.bank = torch.randn(size).type_as(self.bank)
        self.bank = torch.nn.functional.normalize(self.bank, dim=-1)
        self.bank_ptr = torch.zeros(1).type_as(self.bank_ptr)
```

Also, we should initialize the memory bank when the size is passed in the constructor is a tuple.

**Review by philippmwirth — CHANGES_REQUESTED** [review]

File: lightly/models/modules/memory_bank.py (hunk 4 above)
Comment (guarin) [reply lightly/models/modules/memory_bank.py]: Updated :)

> Also, we should initialize the memory bank when the size is passed in the constructor is a tuple.

This is already implemented.

File: benchmarks/imagenet/resnet50/mocov2.py [hunk 5]
```diff
@@ -33,7 +33,11 @@ def __init__(self, batch_size_per_device: int, num_classes: int) -> None:
         self.projection_head = MoCoProjectionHead()
         self.query_backbone = copy.deepcopy(self.backbone)
         self.query_projection_head = MoCoProjectionHead()
-        self.criterion = NTXentLoss(temperature=0.2, memory_bank_size=(65536, 128))
+        self.criterion = NTXentLoss(
+            temperature=0.2,
+            memory_bank_size=(65536, 128),
+            gather_distributed=self.trainer.num_devices > 1,
```
Comment (guarin) [inline benchmarks/imagenet/resnet50/mocov2.py:39]: This makes sure that the memory bank is identical over all processes (keys are gathered before the memory bank is updated). This follows the original implementation: https://github.com/facebookresearch/moco/blob/1999df86de600027cd2702cbd7cd7b7d2a53559d/moco/builder.py#L68
